src/scipp/core/binning.py

```diff
@@ -0,0 +1,111 @@
+    elif isinstance(arg, Variable):
+        if arg.ndim == 0:
+            start = coord.min()
+            step = arg.to(dtype=start.dtype, unit=start.unit)
+            stop = _upper_bound(coord) + step
+            coord = arange(name, start, stop, step=step)
+        else:
+            # TODO check name in arg.dims
+            # TODO handle 2D coords, name must be used, but current implementation
+            # has no support for this
+            coord = arg
+    return coord
+
+
+def _make_edges(x: Union[_cpp.DataArray,
+                         _cpp.Dataset], arg_dict: Dict[str, Union[int, Variable]],
+                kwargs: Dict[str, Union[int, Variable]]) -> List[Variable]:
+    # TODO this merging approach does not make much sense, because the user does not
+    # have control over order. Accept either one or the other.
+    # TODO Could go with kwargs only, user can use **{'name with space':5}
+    if arg_dict is not None:
+        kwargs = dict(**arg_dict, **kwargs)
+    return {name: _parse_coords_arg(x, name, arg) for name, arg in kwargs.items()}
+
+
+def _hist(x: Union[_cpp.DataArray, _cpp.Dataset],
+          arg_dict: Optional[Dict[str, Union[int, Variable]]] = None,
+          /,
+          **kwargs) -> Union[_cpp.DataArray, _cpp.Dataset]:
+    edges = list(_make_edges(x, arg_dict, kwargs).values())
+    if len(edges) == 0:
+        return x.bins.sum()
+    if len(edges) == 1:
+        # TODO Note that this may swap dims, is that ok?
+        return histogram(x, bins=edges[0])
+    else:
+        return histogram(bin(x, edges=edges[:-1]), bins=edges[-1])
+
+
+def _nanhist(x: Union[_cpp.DataArray, _cpp.Dataset],
+             arg_dict: Optional[Dict[str, Union[int, Variable]]] = None,
+             /,
+             **kwargs) -> Union[_cpp.DataArray, _cpp.Dataset]:
+    edges = list(_make_edges(x, arg_dict, kwargs).values())
+    if len(edges) == 0:
+        return x.bins.nansum()
+    return bin(x, edges=edges).bins.nansum()
+
+
+def _bin(x: Union[_cpp.DataArray, _cpp.Dataset],
+         arg_dict: Dict[str, Union[int, Variable]] = None,
+         /,
+         **kwargs) -> Union[_cpp.DataArray, _cpp.Dataset]:
+    edges = list(_make_edges(x, arg_dict, kwargs).values())
+    return bin(x, edges=edges)
+
+
+def _rebin(x: Union[_cpp.DataArray, _cpp.Dataset],
+           arg_dict: Dict[str, Union[int, Variable]] = None,
+           /,
+           **kwargs) -> Union[_cpp.DataArray, _cpp.Dataset]:
+    edges = _make_edges(x, arg_dict, kwargs)
+    out = x
+    for dim, edge in edges.items():
+        out = rebin(out, dim, edge)
+    return out
+
+
+def _make_groups(x, arg):
+    import numpy as np
+    if isinstance(arg, Variable):
+        return arg
+    coord = x.meta.get(arg)
+    # TODO Check that it is not bin-edges?
+    # TODO Very inefficient concat and np.unique
+    if coord is None:
+        return _make_groups(x.data.flatten(to='dummy').bins.concat('dummy').value, arg)
```

Comment:
What does this case do?

Reply:
It handles event coords.

Reply:
So is the purpose to extract an event coord but avoiding the pitfalls of `constituents`? Could you also do `x.bins.coords[arg].copy().bins.constituents['data']`?

Reply:
No, that would be broken when there are gaps between bins, e.g., when we deal with a slice.

Reply:
Yes, copy should work. I'll go with that since it is likely more efficient than `concat`.